ui/src/app/utils/cluster-initializer.tsx

```diff
@@ -0,0 +1,301 @@
+    if (this.abortSignal && this.abortSignal.aborted) {
+      // We'll bail out early if an abort signal was triggered while waiting for the poll cycle.
+      return this.reject(
+        new ClusterInitializationFailedError('Request was aborted.', this.currentCluster));
+    }
+    if (Date.now() - this.initializeStartTime > this.overallTimeout) {
+      return this.reject(
+        new ClusterInitializationFailedError(
+          'Initialization attempt took longer than the max time allowed.',
+          this.currentCluster));
+    }
+
+    // Fetch the current cluster status, with some graceful error handling for NOT_FOUND response
+    // and abort signals.
+    try {
+      this.currentCluster = await this.getCluster();
+      this.onStatusUpdate(this.currentCluster.status);
+    } catch (e) {
+      if (isAbortError(e)) {
+        return this.reject(
+          new ClusterInitializationFailedError('Abort signal received during cluster API call',
+            this.currentCluster));
+      } else if (this.isNotFoundError(e)) {
+        // A not-found error is somewhat expected, if a cluster has recently been deleted or
+        // hasn't been created yet.
+        this.currentCluster = null;
+      } else {
+        this.handleUnknownError(e);
+        if (this.tooManyServerErrors()) {
+          return this.reject(
+            new ExceededErrorCountError('Reached max server error count', this.currentCluster));
+        }
+      }
+    }
+
+    // Attempt to take the appropriate next action given the current cluster status.
+    try {
+      if (this.currentCluster === null) {
+        await this.createCluster();
+      } else if (this.isClusterStopped()) {
+        await this.resumeCluster();
+      } else if (this.isClusterErrored()) {
+        // If cluster is in error state, delete it so it can be re-created at the next poll loop.
+        reportError(
+          `Cluster ${this.currentCluster.clusterNamespace}/${this.currentCluster.clusterName}` +
+          ` has reached an ERROR status.`);
+        await this.deleteCluster();
+      } else if (this.isClusterRunning()) {
+        // We've reached the goal - resolve the Promise.
+        return this.resolve(this.currentCluster);
+      }
+    } catch (e) {
+      if (isAbortError(e)) {
+        return this.reject(
+          new ClusterInitializationFailedError('Abort signal received during cluster API call',
+          this.currentCluster));
+      } else if (e instanceof ExceededActionCountError) {
+        // This is a signal that we should hard-abort the polling loop due to reaching the max
+        // number of delete or create actions allowed.
+        return this.reject(e);
+      } else {
+        this.handleUnknownError(e);
+        if (this.tooManyServerErrors()) {
+          return this.reject(
+            new ExceededErrorCountError('Reached max server error count', this.currentCluster));
+        }
+      }
+    }
+
+    setTimeout(() => this.poll(), this.currentDelay);
```

Comment:
nit: do we already have backoff delay programmed someplace? Seems odd for it to live here.

Reply:
OK, but there's also a lot of maintenance overhead if we have multiple, slightly inconsistent ways of doing basic things like retrying.